src/topp/DatabaseSuitability.cpp

```diff
@@ -0,0 +1,372 @@
+      os << "#ms2_spectra" << count_ms2_lvl << "\n";
+      os << "spectral\t" << id_rate << "\n";
+      os.close();
+    }
+
+    return EXECUTION_OK;
+
+  }
+
+private:
+  double getDecoyDiff_(const PeptideIdentification& pep_id)
+  {
+    double diff = DBL_MAX;
+
+    // get the score of the first two decoy hits
+    double decoy_1 = DBL_MAX;
+    double decoy_2 = DBL_MAX;
+    UInt curr_hit = 0;
+
+    for (const auto& hit : pep_id.getHits())
+    {
+      if (curr_hit > 10) break;
+      ++curr_hit;
+
+      if (!hit.metaValueExists("target_decoy"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No target/decoy information found! Make sure 'PeptideIndexer' is run beforehand."));
+      }
+
+      if (pep_id.getScoreType() != "q-value")
+      {
+        if (!hit.metaValueExists("q-value"))
+        {
+          throw(Exception::Precondition(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No q-value found at peptide identification nor at peptide hits. Make sure 'False Discovery Rate' is run beforehand."));
+        }
+      }
+
+      if (!hit.metaValueExists("MS:1002252"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No cross correlation score found at peptide hit. Only Comet search engine is supported right now."));
+      }
+
+      if (decoy_1 == DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_1 = hit.getMetaValue("MS:1002252");
+        continue;
+      }
+      if (decoy_1 < DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_2 = hit.getMetaValue("MS:1002252");
+        break;
+      }
+    }
+
+    if (decoy_2 < DBL_MAX) // if there are two decoy hits
+    {
+      diff = abs(decoy_1 - decoy_2) / pep_id.getMZ(); // normalized by mw
+    }
+
+    // if there aren't two decoy hits -1 is returned
+    return diff;
+  }
+
+  double getDecoyCutOff_(const vector<PeptideIdentification>& pep_ids, double novor_fract)
+  {
+    // get all decoy diffs of peptide ids with at least two decoy hits
+    vector<double> diffs;
+    for (const auto& pep_id : pep_ids)
+    {
+      double diff = getDecoyDiff_(pep_id);
+      if (diff < DBL_MAX)
+      {
+        diffs.push_back(diff);
+      }
+    }
+
+    if (double(diffs.size()) / pep_ids.size() < 0.2)
+    {
+      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Under 20 % of peptide identifications have two decoy hits. This is not enough for re-ranking. Use the 'force_no_re_rank' flag to still compute a suitability score."));
+    }
+
+    // sort the diffs decreasing and get the (1-novo_fract)*N one
+    auto sort_end = diffs.begin() + (1 - novor_fract) * diffs.size();
+    
+    std::partial_sort(diffs.begin(), sort_end + 1, diffs.end(), std::greater<double>());
+
+    return *sort_end;
+  }
+
+  bool isNovoHit_(const PeptideHit& hit)
+  {
+    set<String> accessions = hit.extractProteinAccessionsSet();
```

Comment:
can be `const set<String> accesssions = `
just to avoid the copy if `hit.extractProteinAccessionsSet()` returns an l-value (in the future)